components/zcash_address/src/kind/unified.rs

```diff
@@ -75,19 +76,25 @@ impl PartialOrd for Typecode {
     }
 }
 
-impl From<u8> for Typecode {
-    fn from(typecode: u8) -> Self {
-        match typecode {
-            0x00 => Typecode::P2pkh,
-            0x01 => Typecode::P2sh,
-            0x02 => Typecode::Sapling,
-            0x03 => Typecode::Orchard,
-            _ => Typecode::Unknown(typecode),
+impl TryFrom<u32> for Typecode {
+    type Error = ParseError;
+
+    fn try_from(typecode: u32) -> Result<Self, Self::Error> {
+        if typecode < 0x02000000 {
+            Ok(match typecode {
+                0x00 => Typecode::P2pkh,
+                0x01 => Typecode::P2sh,
+                0x02 => Typecode::Sapling,
+                0x03 => Typecode::Orchard,
+                _ => Typecode::Unknown(typecode),
+            })
+        } else {
+            Err(ParseError::InvalidTypecodeValue(typecode as u64))
         }
```

Comment:
I would prefer this syntax, which is simpler (and the compiler can reason about) while keeping the invalid range explicit:
```suggestion
        match typecode {
            0x00 => Ok(Typecode::P2pkh),
            0x01 => Ok(Typecode::P2sh),
            0x02 => Ok(Typecode::Sapling),
            0x03 => Ok(Typecode::Orchard),
            0x02000001..=u32::MAX => Err(ParseError::InvalidTypecodeValue(typecode as u64)),
            _ => Ok(Typecode::Unknown(typecode)),
        }
```

Reply:
Or to be even more explicit (and have the compiler check that all values are covered):
```suggestion
        match typecode {
            0x00 => Ok(Typecode::P2pkh),
            0x01 => Ok(Typecode::P2sh),
            0x02 => Ok(Typecode::Sapling),
            0x03 => Ok(Typecode::Orchard),
            0x04..=0x02000000 => Ok(Typecode::Unknown(typecode)),
            0x02000001..=u32::MAX => Err(ParseError::InvalidTypecodeValue(typecode as u64)),
        }
```